src/Runner/Example.hs

```diff
@@ -13,52 +13,118 @@ data Result = Equal | NotEqual [String]
   deriving (Eq, Show)
 
 mkResult :: ExpectedResult -> [String] -> Result
-mkResult expected actual
-  | expected `matches` actual = Equal
-  | otherwise = NotEqual (formatNotEqual expected actual)
+mkResult expected_ actual_ =
+  case expected `matches` actual of
+  Nothing         -> Equal
+  Just (col, row) -> NotEqual (formatNotEqual expected actual col row)
   where
-    chunksMatch :: [LineChunk] -> String -> Bool
-    chunksMatch [] "" = True
-    chunksMatch [LineChunk xs] ys = stripEnd xs == stripEnd ys
+    -- use show to escape special characters in output lines if any output line
+    -- contains any unsafe character
+    escapeOutput
+      | any (not . isSafe) $ concat (expectedAsString ++ actual_) = init . tail . show . stripEnd
+      | otherwise = id
+
+    actual :: [String]
+    actual = escapeOutput <$> actual_
+
+    expected :: ExpectedResult
+    expected = transformExcpectedLine escapeOutput <$> expected_
+
+    expectedAsString :: [String]
+    expectedAsString = map (\x -> case x of
+        ExpectedLine str -> concatMap lineChunkToString str
+        WildCardLine -> "..." ) expected_
+
+    isSafe :: Char -> Bool
+    isSafe c = c == ' ' || (isPrint c && (not . isSpace) c)
+
+    chunksMatch :: [LineChunk] -> String -> Maybe (String, String)
+    chunksMatch [] "" = Nothing
+    chunksMatch [LineChunk xs] ys =
+      if stripEnd xs == stripEnd ys
+      then Nothing
+      else matchingPrefix xs ys
     chunksMatch (LineChunk x : xs) ys =
-        x `isPrefixOf` ys && xs `chunksMatch` drop (length x) ys
+      if x `isPrefixOf` ys
+      then (\(m, p) -> (x ++ m, x ++ p)) <$> (xs `chunksMatch` drop (length x) ys)
+      else matchingPrefix x ys
     chunksMatch zs@(WildCardChunk : xs) (_:ys) =
-        xs `chunksMatch` ys || zs `chunksMatch` ys
-    chunksMatch _ _ = False
+      case xs `chunksMatch` ys of
+      Nothing  -> Nothing
+      Just (m, col) -> case zs `chunksMatch` ys of
+                       Nothing   -> Nothing
+                       Just (m2, col2) ->
+                         let (mm, coll) = if length m2 > length m then (m2, col2) else (m, col) in
+                         Just $ (mm, '.':coll)
```

Comment:
Why a full stop? I don't think I understand what purpose the fields are serving. They seem to be using `String` as a unary number.